### ISSUE TITLE
Allow to add mandatory parameters for Client constructors

### ISSUE THREAD

**Issue by yangyuan** [Azure/autorest#2962]
Background: Cognitive Services have multiple possible endpoints, basically it's like ```http[s]://{anydomain}[:{anyport}]/{anybasepath}```

Right now, we are using host and a placeholder uri:
```host: https://placeholder.com/root```
We can also use this:
```x-ms-parameterized-host: {Endpoint}```

But the problem is, for both, the C# Client constructors will make `BaseUri` or `Endpoint` optional.
For example, users can do both:
```
var client = new XxxClient();
client.Endpoint = new Uri("http://10.0.0.4:5000/");
```
```
var client = new XxxClient(new Uri("http://10.0.0.4:5000/"));
```
We want expose the constructors which contain mandatory parameter, so that Client user won't mistakenly forgot to change the default placeholder.

Maybe we can support a new `x-ms-parameter-location`, so the parameter will be in constructor
"x-ms-parameter-location": "client-constructor",

**Comment by fearthecowboy** [comment]
No, this isn't the kind of thing we'd introduce a whole feature for; it's just to specific to a given situation.

To add a feature to make this applicable across all the languages (and then we're stuck supporting it ever forward) would be an incredible cost and testing burden.  

The whole reason we have transform directives is to tweak generated code after it's created so that we can have flexibility in the process.

**Comment by bolig48** [comment]
It makes very much sense to make the URL parameterized so we can choose whether to hardcode it,, read it from a configuration file etc. It is the first question you ask yourself after having generated the client (in my opinion)

By adding it directly into the generator, we ensure it be "less hacky" and that it will work in future releases. How the generator chooses to implement it I am less concerned about. Perhaps you can simply use main elements from the above proposed strategy.

**Comment by brjohnstmsft** [comment]
@fearthecowboy In general, why should we use transform directives to solve this class of problem instead of, say, partial classes in C#? It doesn't seem to add any value over partial classes; In fact, it adds even more complexity since we're encoding code in a specific target language in JS (and need regexes too).
